dmcontent/questions.py

```diff
@@ -480,6 +479,51 @@ def summary(self, service_data):
         return ListSummary(self, service_data)
 
 
+class Hierarchy(List):
+    """
+    For our purposes, a Hierarchy is like a List, except the entries
+    are potentially related to each other in a "subsumptive
+    containment hierarchy". For example, every service that is a member
+    of a child category is _by definition_ also a member of its parent
+    category as well, and we can add those parent categories as a
+    denormalization to make searching more efficient.
+    """
+
+    def _get_data(self, form_data):
+        if self.id not in form_data:
+            return {self.id: None}
+
+        values = set(form_data.getlist(self.id))
+        values.update(self.get_missing_values(values))
+
+        return {self.id: sorted(values) or None}
+
+    def summary(self, service_data):
+        return HierarchySummary(self, service_data)
+
+    def get_missing_values(self, selected_values_set):
+        """
+        Recursively retrieves un-selected parent categories of the
+        passed-in selection, as a set of 'value' strings.
+        :param selected_values_set: initially-selected categories (e.g. by the user)
+        :return: additional values that should also be selected
+        """
+
+        def update_expected_values(options, parents, expected_set):
```

Comment:
@allait is this ok now? - didn't want to make it return, as it felt confusing to both return _and_ update the passed-in object, and I don't want to make copies of the set everywhere (which I feel would be implied by having the return)